### PR TITLE
Pass extensions to node if extension processing is handled by it

### PR DESCRIPTION
**Pull request by hpatro**
Ref: https://github.com/redis/redis/pull/12760

#### Fixes compatibilty of PlaceholderKV cluster (7.2 - extensions enabled by default) with older Redis cluster (< 7.0 - extensions not handled) .

With some of the extensions enabled by default in 7.2 version, new nodes running 7.2 and above start sending out larger clusterbus message payload including the ping extensions. This caused an incompatibility with node running engine versions < 7.0. Old nodes (< 7.0) would receive the payload from new nodes (> 7.2) would observe a payload length (totlen) > (estlen) and would perform an early exit and won't process the message.

Note: A successful `PING`/`PONG` is required as a sender for a given node to be marked as `extensions_supported` and then extensions message will be sent to it. This could cause a slight delay in receiving the extensions message(s). 

### Testing

TCL test verifying the cluster state is healthy irrespective of enabling/disabling cluster message extensions feature.

### CODE REVIEW

**Comment by hpatro** [comment]
One of the test fails in few of the runs. 
```
[err]: Verify the nodes configured with prefer hostname only show hostname for new nodes in tests/unit/cluster/hostnames.tcl
Expected '' to be equal to 'shard-1.com' (context: type eval line 60 cmd {assert_equal [lindex [get_slot_field $slot_result 0 2 3] 1] "shard-1.com"} proc ::test)
```
Taking a look.

**Comment by hpatro** [comment]
Regarding the flaky test. There is a slight behavior change with this PR.

A successful PING/PONG is required as a sender for a given node to be marked as extensions_supported and then extensions message will be sent to it. This could cause a slight delay in receiving the extensions message(s). (Added it to the top level comment as well)

**Comment by madolson** [comment]
> A successful PING/PONG is required as a sender for a given node to be marked as extensions_supported and then extensions message will be sent to it. This could cause a slight delay in receiving the extensions message(s). (Added it to the top level comment as well)

So this happens in two flows:
1. The initial meet flow, where a new node to a cluster has to omit extensions since it doesn't know if the target supports them. After the target of the meet receives the message, it will show the node without extensions (not great) until it get's a followup pong message. We could solve this by omitting a node from cluster output (if it sent us a message with the extension bit set).
2. When gossiping, other nodes will learn about the new node, and they also send it a message without extensions. This results as the same problem as one, but is easier to resolve since we can gossip if a node supports extensions. There ate 16 bits of flags (of which only 10 are being used) on the node flags that 

So we really just need a goos solution for 1.

File: src/cluster_legacy.c
Comment (PingXie) [inline src/cluster_legacy.c]: We should set `CLUSTERMSG_FLAG0_NODE_SUPPORTS_EXTENSIONS` in `writePingExt()`, which is the reason why this flag needs to be sent. Also it is more logical to set this flag only when the receiver supports extension.

File: src/cluster_legacy.h
Comment (PingXie) [inline src/cluster_legacy.h]:
```suggestion
    bool extensions_supported; /* Does the node support parsing ping extensions. */
```

File: src/cluster_legacy.c
Comment (PingXie) [inline src/cluster_legacy.c]: This is a bit hacky IMO. Rather, I would suggest pulling L2617 (or L2624) up and bailing out of this function if `extensions == 0`, as a more generic fix.

File: src/cluster_legacy.c
Comment (PingXie) [inline src/cluster_legacy.c]: I don't think `server.process_clustermsg_extensions` is needed. The intent here is to emulate a non-extension capable **sender**. It is actually a good idea to harden the receiver logic so it could deal with such a sender. We should revert this line.

File: src/cluster_legacy.h
Comment (PingXie) [reply src/cluster_legacy.h]: On a second thought, should we make this a flag instead?

File: src/cluster_legacy.c
Comment (PingXie) [inline src/cluster_legacy.c]: existing 7.2/7.0 will not carry `CLUSTERMSG_FLAG0_NODE_SUPPORTS_EXTENSIONS` hence they would be marked as "not capable of supporting extensions", which is not accurate. Shouldn't we check `hdr->mflags[0] & CLUSTERMSG_FLAG0_EXT_DATA` instead?

File: src/cluster_legacy.h
Comment (PingXie) [inline src/cluster_legacy.h]: Now I am not sure if we even need this flag - I think it is a dup of `CLUSTERMSG_FLAG0_EXT_DATA`.

File: src/cluster_legacy.c
Comment (PingXie) [inline src/cluster_legacy.c]: This debug flag is a bit hacky IMO as it changes the exact contract of the function. I understand this is a smaller change, but I feel it would be cleaner if this debug flag can be evaluated on the caller side.

File: src/server.h
Comment (PingXie) [inline src/server.h]: I know we don't have this practice yet but can we consider using a common prefix for all debug flags? such as `dbg_`?

also if we make the receiver agnostic to the extension and only use this flag to control the sender's behavior, we should probably consider a different name, such as `dbg_node_has_no_extension` or something along that line?

Comment (daniel-house) [reply src/server.h]: Or perhaps put all debug flags in a sub-structure.

File: src/cluster_legacy.h
Comment (madolson) [reply src/cluster_legacy.h]: Yes, on the redis issue I suggested it be added into the flags. That also makes it automatically get propagated in the gossip.

File: src/server.h
Comment (madolson) [reply src/server.h]: A strong opinion from me, I want to stop adding them to the debug command and instead actually just build normal configs optionally for debug builds that aren't in the release build.

File: src/cluster_legacy.h
Comment (madolson) [reply src/cluster_legacy.h]: No, it's not the same. We need a way to indicate we are sending extensions without sending the extensions. Otherwise it'll get evaluated as if this node has zero extensions, which breaks the logic for hostnames and the TLS messages, since we assume if a node doesn't send them it removed them.

File: src/cluster_legacy.c
Comment (madolson) [reply src/cluster_legacy.c]: Ideally we would have intra-version tests (or unit tests) as well.

**Comment by hpatro** [comment]
@PingXie @madolson I'm addressing the comments on the PR. Will get back in couple of days.

**Comment by hpatro** [comment]
@zuiderkwast Unsure about what you meant by assigning it to Ping for this PR, did you meant it for review or something else ?

File: src/cluster_legacy.h
Comment (hpatro) [reply src/cluster_legacy.h]: I think I agree with you @PingXie . @madolson and I discussed a bit more on this and the problem (display nodes only on receiving complete information) we're trying to solve seems not practical without patching older version(s). Will incorporate this feedback.

**Comment by PingXie** [comment]
> @zuiderkwast Unsure about what you meant by assigning it to Ping for this PR, did you meant it for review or something else ?

@hpatro please continue with this PR. I will be the "designated" reviewer :-)

**Comment by zuiderkwast** [comment]
Yeah, I meant as the reviewer. Sorry.

**Comment by zuiderkwast** [comment]
Question: Shall we include this in our 7.2.4 release?

Although it improves compatibility with older Redis versions, it's a difference compared to Redis 7.2.4. We should mark this clearly in our release notes then.

**Comment by PingXie** [comment]
> Shall we include this in our 7.2.4 release?

I think we should try our best to make it happen. It is an important and **fully compatible** fix.

File: src/cluster_legacy.c
Comment (hpatro) [reply src/cluster_legacy.c]: Thanks for your suggestion. Getting rid of this flag and using `CLUSTERMSG_FLAG0_EXT_DATA`.

File: src/server.h
Comment (hpatro) [reply src/server.h]: `enable-debug-command` is a good enough guardrail at this moment and I feel it doesn't need to be addressed with this PR. I'm renaming it to `dbg_node_send_extensions` defaults to `1`. WDYT @madolson @PingXie ?

**Comment by madolson** [comment]
> @hpatro please continue with this PR. I will be the "designated" reviewer :-)

Reminds me of the fact we should add code owner when we get this all sorted.

**Comment by madolson** [comment]
> I think we should try our best to make it happen. It is an important and fully compatible fix.

Yeah, I was pushing for this because I think it's important for compatibility. I would be okay with this in like a 7.2.5 though, it's important to have eventually.

File: src/cluster_legacy.c
Comment (hpatro) [reply src/cluster_legacy.c]: @PingXie Fair enough, there are not enough caller side touch point, I'm fine moving it there.

@madolson Yes, we should build it :) .

File: src/server.h
Comment (hpatro) [reply src/server.h]: @daniel-house Yeah, we could do that as well. Leaving it around the top level struct for now.

**Comment by hpatro** [comment]
I think I can submit a PR today. And get it done by Monday if you all can help review it @madolson @PingXie . It would be nice to have this in our first release itself.

**Comment by hpatro** [comment]
Changes done:

1. Introduce new node flag `CLUSTER_NODE_EXTENSIONS_SUPPORTED` to mark if the node can handle extension(s) data based on `CLUSTERMSG_FLAG0_EXT_DATA` cluster message header flag.
2. Send extensions data only to the node which can handle extensions based on pt. 1.
3. `CLUSTERMSG_FLAG0_EXT_DATA` header flag would be sent irrespective of extensions present or not. This helps indicate if the node supports extensions feature or not.
4. Introduce debug flag to enable/disable sending extensions data. Receiving extensions is still enabled.

@PingXie @madolson Let me know if you agree with this.


@madolson Regarding your concern of displaying IP address incorrectly for a period of time, I think we should decouple it from this issue. And I think that issue was valid in 7.0 itself. Can discuss further.

File: src/cluster_legacy.c
Comment (PingXie) [inline src/cluster_legacy.c]: nit - this would read as if we only send extensions when the dbg flag is enabled. let's use the invert instead.

```suggestion
    if (!server.dbg_suppress_clustermsg_extensions && link->node && nodeSupportsExtensions(link->node)) {
        estlen += writePingExt(NULL, 0);
    }
```

File: src/cluster_legacy.c
Comment (PingXie) [inline src/cluster_legacy.c]:
```suggestion
    if (!server.dbg_suppress_clustermsg_extensions && link->node && nodeSupportsExtensions(link->node)) {
```

File: src/debug.c
Comment (PingXie) [inline src/debug.c]:
```suggestion
    } else if (!strcasecmp(c->argv[1]->ptr,"suppress-clustermsg-extensions") && c->argc == 3) {
```

File: src/debug.c
Comment (PingXie) [inline src/debug.c]:
```suggestion
"SUPPRESS-CLUSTERMSG-EXTENSIONS <0|1>",
"    Set to 1 to not broadcast cluster message extensions.",
```

File: src/debug.c
Comment (PingXie) [inline src/debug.c]:
```suggestion
        server.dbg_suppress_clustermsg_extensions = atoi(c->argv[2]->ptr);
```

File: src/server.c
Comment (PingXie) [inline src/server.c]:
```suggestion
    server.dbg_suppress_clustermsg_extensions = 0;
```

Comment (PingXie) [reply src/server.c]: also, I think we should move all dbg_* settings into its own "sections" and not mixed normal server settings

File: src/server.h
Comment (PingXie) [inline src/server.h]:
```suggestion
    int dbg_suppress_clustermsg_extensions; /* Debug config that when set suppresses  aux cluster message extensions broadcasting. */
```

File: tests/unit/cluster/extensions-parsing.tcl
Comment (PingXie) [inline tests/unit/cluster/extensions-parsing.tcl]: This test is not 100% correct. What we want is a permanent state where only `node-1` knows about `host-1.com`. What this test validates is that, at one point, there is one node, not necessarily  `node-1`, knows about `host-1.com`. Even without the test hook setup on L33-34, this check could pass, depending on timing. so there is a false negative bug in this logic.

File: src/cluster_legacy.c
Comment (PingXie) [reply src/cluster_legacy.c]: on a second thought, this dbg flag doesn't do anything because our receiver knows how to handle the extensions, now that we no longer disable extension handling logic using the dbg flag, which I think is the right thing to do.

I think a lot of the complexity of this fix comes from trying to create a TCL test for ad hoc version upgrade tests.  This is going to create headaches in the long run. I would suggest removing this debug flag and the TCL tests, and then validating the upgrade path *manually* for this fix. Let's defer discussion of the version upgrade test framework.

**Comment by hpatro** [comment]
@PingXie Thanks for your feedback. 

I initially started off this PR (in Redis) with similar setup you suggested (fixing the issue and manually tested it). The fake testing approach seems to bring in a bit of complexity and albeit it wasn't perfect (tried to get it out in a hurry). I've trimmed down the code, please have a look now.

**Comment by hpatro** [comment]
Tested few things 

### Setup
* 2 cluster (1 primary/1 replica) each 
* Shard 1: Engine Redis 6.2
* Shard 2: Engine Valkey latest with patch

### Operations
*  MEET between the hybrid cluster - Cluster was stable during the operation and slot information got propagated correctly between the clusters.
* Replaced primary of Redis 6.2 with Valkey latest, cluster/node remained healthy after the replacement. 
* Added a Redis 7.2 node (without the fix), shows up as `fail` for the Redis 6.2 and remains connected to the Valkey latest node.

File: tests/unit/cluster/hostnames.tcl
Comment (PingXie) [inline tests/unit/cluster/hostnames.tcl]: I feel that it would be better to keep the original L173-L176 and insert a `wait_for_cluster_propagation` call right before the checks.

Also, I think there is similar flakiness on L153-L155.

**Review by PingXie — COMMENTED** [review]
This looks a lot cleaner and maintainable to me. Thanks a lot for simplifying the fix, @hpatro! 

The server code LGTM but I think there is room for improvement in the test code. Can you please take another look?

File: tests/unit/cluster/hostnames.tcl
Comment (hpatro) [reply tests/unit/cluster/hostnames.tcl]: I've applied a fix around L153-155, WDYT about it, I'm thinking of applying it similarly on L173-L176 to avoid flakiness.

File: tests/unit/cluster/hostnames.tcl
Comment (PingXie) [inline tests/unit/cluster/hostnames.tcl:120]: Super nit - starting `j` from `1` would avoid the changes on L154 and L160. It took me a while to understand the shift of the shard "index". 
```suggestion
    for {set j 1} {$j < $primaries} {incr j} {
```

**Review by PingXie — APPROVED** [review]
LGTM

**Comment by madolson** [comment]
> @madolson Regarding your concern of displaying IP address incorrectly for a period of time, I think we should decouple it from this issue. And I think that issue was valid in 7.0 itself. Can discuss further.

Why do you think this? I think this is a regression if you have hostnames enabled and have a new node join a cluster, as for a brief period of time the nodes will incorrect show up as not support hostnames (even though they do). We've seen issues in the past where clients don't handle this correctly, and some clients see prolonged periods without being able to resolve it. My concern is that it affects new nodes added during steady state, and not just the upgrade case.

**Comment by hpatro** [comment]
> > @madolson Regarding your concern of displaying IP address incorrectly for a period of time, I think we should decouple it from this issue. And I think that issue was valid in 7.0 itself. Can discuss further.
> 
> 
> 
> Why do you think this? I think this is a regression if you have hostnames enabled and have a new node join a cluster, as for a brief period of time the nodes will incorrect show up as not support hostnames (even though they do). We've seen issues in the past where clients don't handle this correctly, and some clients see prolonged periods without being able to resolve it. My concern is that it affects new nodes added during steady state, and not just the upgrade case.

Exactly, this issue exists for a new node being added and not related to the upgrade scenario. So, it could happen for any cluster running with hostnames feature. 
Hence, I don't want to couple with this particular PR.

**Comment by madolson** [comment]
> Exactly, this issue exists for a new node being added and not related to the upgrade scenario. So, it could happen for any cluster running with hostnames feature.

Today, the hostname feature blindly sends the extension when it joins a new cluster, assuming it supports it. Now, with this change, we will wait until we have confirmation it supports the hostname. This is different behavior no?

**Comment by hpatro** [comment]
> > Exactly, this issue exists for a new node being added and not related to the upgrade scenario. So, it could happen for any cluster running with hostnames feature.
> 
> 
> 
> Today, the hostname feature blindly sends the extension when it joins a new cluster, assuming it supports it. Now, with this change, we will wait until we have confirmation it supports the hostname. This is different behavior no?

Won't that node be added via gossip as well and in the gossip information, hostname won't be present. Am I missing something? 🙈

**Comment by madolson** [comment]
> Won't that node be added via gossip as well and in the gossip information, hostname won't be present. Am I missing something? 🙈

That was one of the two cases we discussed, but the case I'm talking about is for meet messages. The first message a node sends to a new cluster will not include its hostname. (Maybe I'm indexing too much on this case though).

**Comment by PingXie** [comment]
> > Exactly, this issue exists for a new node being added and not related to the upgrade scenario. So, it could happen for any cluster running with hostnames feature.
> 
> 
> 
> Today, the hostname feature blindly sends the extension when it joins a new cluster, assuming it supports it. Now, with this change, we will wait until we have confirmation it supports the hostname. This is different behavior no?

Why is this a problem? The cluster bus is an eventual consistency system, and the information broadcast via the cluster bus including "hostname" is inherently dynamic/non-constant. so conceptually speaking this behavior change is within the design premise.

**Comment by madolson** [comment]
> The cluster bus is an eventual consistency system, and the information broadcast via the cluster bus including "hostname" is inherently dynamic/non-constant.

Although generally this is true, in practice today if you set all the hostnames at startup a node will never observe a node that doesn't have a hostname. This change introduces a small window where that isn't true, and it could show it to clients and a user might see errors. 

I suppose I need to be prescriptive. It seems like everyone wants something simple, so I can lean into that. I would rather hostnames continue to work as they do, with nodes not checking to see if the other node supports it.

File: tests/unit/cluster/hostnames.tcl
Comment (hpatro) [reply tests/unit/cluster/hostnames.tcl:120]: The server nodes are 0-indexed so we would have to perform an expression to apply the 1-based shard id appropriately. Hence, I think the current change is more simpler.  

0th node -> shard-0.com
1st node -> shard-1.com and so on.

**Review by madolson — APPROVED** [review]
Besides my comment I was OK with the design, so would rather have this in the RC candidate. Folks can comment.

**Comment by hpatro** [comment]
> Besides my comment I was OK with the design, so would rather have this in the RC candidate. Folks can comment.

Will file an issue explaining the scenarios better and we can decide over there.

**Comment by PingXie** [comment]
> if you set all the hostnames at startup a node will never observe a node that doesn't have a hostname. This change introduces a small window where that isn't true, and it could show it to clients and a user might see errors.

First and foremost, I am not sure about the importance/value of this property  of "a node will never observe a node that doesn't have a hostname". This already can happen during rolling upgrades from pre 7.0 builds to 7.2 and all nodes will have to handle it. 

Furthermore, the propagation of many other node attributes don't have this property either today. Even within the same node, the cluster and the replication modules might have a different view on the "primary-replica" relationship too (for a very short period of time after `cluster replicate`). 

My point is that, with the current cluster design, the operator would have to perform an explicit synchronization regardless (think of `wait_for_cluster_propagation`). So a one-off effort to provide this property (of "a node will never observe a node that doesn't have a hostname") is not going to change the landscape much but with the downside of the excessive complexity (to a part of the system where we already are grappling with reliability issues).

**Comment by madolson** [comment]
> First and foremost, I am not sure about the importance/value of this property of "a node will never observe a node that doesn't have a hostname". This already can happen during rolling upgrades from pre 7.0 builds to 7.2 and all nodes will have to handle it.

The important point here is a node can't have a hostname pre 7.0, it was an opt-in feature.

**Pull request merged**
